### ISSUE TITLE
Why are you not "returning" the updated state in the state monad?

### ISSUE THREAD

**Issue by endrebak** [dmvianna/haskellbook#1]
Your solution is 

``` haskell
 (>>=) :: Moi s a
        -> (a -> Moi s b)
        -> Moi s b
  (Moi f) >>= g = Moi $ \s -> let a = fst $ f s
                                  ms = runMoi $ g a
                              in ms s
```

while mine is 

``` haskell
  (Moi f) >>= g = Moi $ \s -> let (a, s') = f s
                                  sb = g a
                                  (b', _) = runMoi sb s
                                  in (b', s')
```

It seems you are actively ignoring the new state by using `fst`. What is your reasoning? I think the monad is for chaining implicit changing state so I suspect your solution is wrong, but I am not confident of it.

**Comment by pasunboneleve** [comment]
Mmm. Let's divide the problem:
1. Both solutions pass the compiler. The main concern when I was solving it was that I could get a `(b, s)`, which I did, and so did you;
2. So we diverge on which `s` we should return. 

I believe I should return the state that `g a` returns. Simple. You think you should return `f s` state. So you artificially pair `b'` to a state that does not correspond to it. 

I don't like your solution. My problem with it is that you can't use that state as a breadcrumb to find your way back to your returned `b`. It isn't a natural pairing, it is a graft.

As I understand it, I'm not chaining states here. I'm chaining values. Were I interested in chaining states, `g` would need to be applied on a state `s`. Instead, it only accepts a value `a` as argument.

**Comment by jjoekoullas** [comment]
Both solutions also pass monad tests from checkers as well, so both agree like that too. :+1:

Here's my implementation

``` haskell
instance Monad (Moi s) where
  return = pure
  (Moi f) >>= g = Moi $ \s ->
    let (a, s')   = f s
        (Moi sb) = g a
    in sb s'
```

The other two solutions don't thread `s` and it's permutations throughout.  The first drops the `s` result from `f s`.  The second drops the `s` result from `runMoi sb s`.  I have a hunch that a state monad would require passing these `s` permutations throughout.

I looked around looking for an implementation of the State monad to see if I could validate my hunch.  I found the [monads chapter from lyah](http://learnyouahaskell.com/for-a-few-monads-more), if you search for "instance Monad (State s) where" you'll find what I'm talking about.

So I don't think either of the other implementations can be considered analogous to the State monad, which was the goal of the exercise.
